File: app/src/main/java/com/zibete/proyecto1/utils/CropHelper.java
// com/zibete/proyecto1/utils/CropHelper.java
package com.zibete.proyecto1.utils;

import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.util.DisplayMetrics;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import androidx.activity.result.ActivityResultLauncher;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.canhub.cropper.CropImageContract;
import com.canhub.cropper.CropImageContractOptions;
import com.canhub.cropper.CropImageOptions;
import com.canhub.cropper.CropImageView;
import com.google.android.gms.tasks.Task;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

public class CropHelper {

    // ====== Fragment ======
    public static ActivityResultLauncher<CropImageContractOptions> registerLauncher(
            Fragment fragment,
            StorageReference refSendImages,
            LinearLayout linear_photo_view,
            LinearLayout linear_photo,
            ImageView photo,
            ProgressBar loadingPhoto,
            android.widget.EditText msg,
            ImageView btnCamera,
            ImageView btnSendMsg,
            java.util.function.Consumer<Uri> onCroppedUri // callback si querés usar la Uri luego
    ) {
        return fragment.registerForActivityResult(new CropImageContract(), result -> {
            if (result.isSuccessful()) {
                Uri resultUri = result.getUriContent();
                applyUiAndUpload(fragment.requireContext(), (AppCompatActivity) fragment.requireActivity(),
                        resultUri, refSendImages, linear_photo_view, linear_photo,
                        photo, loadingPhoto, msg, btnCamera, btnSendMsg);
                if (onCroppedUri != null) if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    onCroppedUri.accept(resultUri);
                }
            } else {
                // manejar error si querés: result.getError()
            }
        });
    }

    // ====== Activity ======
    public static ActivityResultLauncher<CropImageContractOptions> registerLauncherForActivity(
            AppCompatActivity activity,
            StorageReference refSendImages,
            LinearLayout linear_photo_view,
            LinearLayout linear_photo,
            ImageView photo,
            ProgressBar loadingPhoto,
            android.widget.EditText msg,
            ImageView btnCamera,
            ImageView btnSendMsg,
            java.util.function.Consumer<Uri> onCroppedUri
    ) {
        return activity.registerForActivityResult(new CropImageContract(), result -> {
            if (result.isSuccessful()) {
                Uri resultUri = result.getUriContent();
                applyUiAndUpload(activity, activity,
                        resultUri, refSendImages, linear_photo_view, linear_photo,
                        photo, loadingPhoto, msg, btnCamera, btnSendMsg);
                if (onCroppedUri != null) if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    onCroppedUri.accept(resultUri);
                }
            } else {
                // manejar error si querés
            }
        });
    }

    // === común a ambos ===
    private static void applyUiAndUpload(
            Context ctx,
            AppCompatActivity activity,
            Uri resultUri,
            StorageReference refSendImages,
            LinearLayout linear_photo_view,
            LinearLayout linear_photo,
            ImageView photo,
            ProgressBar loadingPhoto,
            android.widget.EditText msg,
            ImageView btnCamera,
            ImageView btnSendMsg
    ) {
        // mostrar UI de “foto lista para enviar”
        linear_photo_view.setVisibility(android.view.View.VISIBLE);
        msg.setVisibility(android.view.View.GONE);
        btnCamera.setVisibility(android.view.View.GONE);
        btnSendMsg.setVisibility(android.view.View.VISIBLE);

        // tamaño del preview
        DisplayMetrics metrics = new DisplayMetrics();
        activity.getWindowManager().getDefaultDisplay().getMetrics(metrics);
        int widthPixels = metrics.widthPixels;
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(widthPixels / 3, widthPixels / 3);
        linear_photo.setLayoutParams(params);

        // subir a Firebase
        String childName = resultUri.getLastPathSegment() != null
                ? resultUri.getLastPathSegment()
                : ("img_" + System.currentTimeMillis());

        StorageReference refPic = refSendImages.child(childName);

        refPic.putFile(resultUri).addOnSuccessListener((UploadTask.TaskSnapshot snap) -> {
            Task<Uri> urlTask = snap.getStorage().getDownloadUrl();
            urlTask.addOnSuccessListener(downloadUri -> {
                loadingPhoto.setVisibility(android.view.View.VISIBLE);
                Glide.with(ctx)
                        .load(downloadUri.toString())
                        .apply(new RequestOptions().transform(new CenterCrop(), new RoundedCorners(35)))
                        .listener(new RequestListener<android.graphics.drawable.Drawable>() {
                            @Override
                            public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<android.graphics.drawable.Drawable> target, boolean isFirstResource) {
                                loadingPhoto.setVisibility(android.view.View.GONE);
                                return false;
                            }

                            @Override
                            public boolean onResourceReady(android.graphics.drawable.Drawable resource, Object model, Target<android.graphics.drawable.Drawable> target, DataSource dataSource, boolean isFirstResource) {
                                loadingPhoto.setVisibility(android.view.View.GONE);
                                return false;
                            }
                        })
                        .into(photo);
            });
        });
    }

    // Lanzar el cropper con opciones comunes
    public static void launchCrop(ActivityResultLauncher<CropImageContractOptions> launcher, @Nullable Uri imageUri) {
        CropImageOptions options = new CropImageOptions();
        options.guidelines = CropImageView.Guidelines.ON;
        options.outputRequestWidth = 1920;
        options.outputRequestHeight = 1080;
        CropImageContractOptions contract = new CropImageContractOptions(imageUri, options);
        launcher.launch(contract);
    }
}

File: app/src/main/java/com/zibete/proyecto1/utils/CropHelper.kt
package com.zibete.proyecto1.utils

import android.content.Context
import android.graphics.drawable.Drawable
import android.net.Uri
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import androidx.activity.result.ActivityResultCaller
import androidx.activity.result.ActivityResultLauncher
import androidx.core.view.isVisible
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.canhub.cropper.CropImageContract
import com.canhub.cropper.CropImageContractOptions
import com.canhub.cropper.CropImageOptions
import com.canhub.cropper.CropImageView
import com.google.firebase.storage.StorageReference

object CropHelper {

    /**
     * Registra el cropper (sirve para Fragment o Activity) y devuelve el launcher.
     *
     * @param caller Fragment/Activity que implementa ActivityResultCaller (p.ej. this)
     * @param ctx Context para recursos/Glide
     * @param refSendImages StorageReference a la carpeta de imágenes donde se sube el recorte
     * @param linearPhotoView Contenedor de la previsualización (se muestra al recortar)
     * @param linearPhoto Vista cuyo tamaño ajustamos para el preview cuadrado
     * @param photo ImageView donde se carga el resultado
     * @param loadingPhoto ProgressBar de carga del preview
     * @param msg EditText del mensaje (se oculta al mostrar foto)
     * @param btnCamera Botón de cámara (se oculta al mostrar foto)
     * @param btnSendMsg Botón de enviar (se muestra al recortar)
     * @param onCroppedUri Callback opcional con la URI recortada (después de subir y previsualizar)
     */
    fun registerLauncher(
        caller: ActivityResultCaller,
        ctx: Context,
        refSendImages: StorageReference,
        linearPhotoView: LinearLayout,
        linearPhoto: LinearLayout,
        photo: ImageView,
        loadingPhoto: ProgressBar,
        loadingButton: ProgressBar,
        frameSendMsg: FrameLayout,
        msg: EditText,
        btnCamera: ImageView,
        btnSendMsg: ImageView,
        onCroppedUri: ((Uri) -> Unit)? = null
    ): ActivityResultLauncher<CropImageContractOptions> {

        return caller.registerForActivityResult(
            CropImageContract()
        ) { result ->
            if (result == null || !result.isSuccessful) {
                // Si querés, log o UI de error con result?.error
                return@registerForActivityResult
            }

            val resultUri = result.uriContent ?: return@registerForActivityResult

            // 1) Mostrar UI de “foto lista para enviar”
            frameSendMsg.isVisible = true
            btnSendMsg.isVisible = false
            msg.isVisible = false
            btnCamera.isVisible = false

            linearPhotoView.isVisible = true
            loadingPhoto.isVisible = true
            loadingButton.isVisible = true

            // 2) Ajustar tamaño de preview (cuadrado = 1/3 del ancho)
            val width = ctx.resources.displayMetrics.widthPixels
            val params = LinearLayout.LayoutParams(width / 3, width / 3)
            linearPhoto.layoutParams = params

            // 3) Subir a Firebase y previsualizar con Glide
            val childName = resultUri.lastPathSegment?.ifBlank { null }
                ?: "img_${System.currentTimeMillis()}"
            val refPic = refSendImages.child(childName)

            refPic.putFile(resultUri)
                .addOnSuccessListener { snap ->
                    snap.storage.downloadUrl.addOnSuccessListener { downloadUri ->
                        // Mostrar spinner mientras carga Glide
                        Glide.with(ctx)
                            .load(downloadUri.toString())
                            .apply(RequestOptions().transform(CenterCrop(), RoundedCorners(35)))
                            .listener(object : RequestListener<Drawable> {
                                override fun onLoadFailed(
                                    e: GlideException?,
                                    model: Any?,
                                    target: Target<Drawable>?,
                                    isFirstResource: Boolean
                                ): Boolean {
                                    return false
                                }

                                override fun onResourceReady(
                                    resource: Drawable?,
                                    model: Any?,
                                    target: Target<Drawable>?,
                                    dataSource: DataSource?,
                                    isFirstResource: Boolean
                                ): Boolean {
                                    return false
                                }
                            })
                            .into(photo)

                        onCroppedUri?.invoke(resultUri)
                    }
                }
        }
    }

    @JvmStatic
    fun launchCrop(
        launcher: ActivityResultLauncher<CropImageContractOptions>,
        imageUri: Uri
    ) {
        val options = CropImageOptions().apply {
            guidelines = CropImageView.Guidelines.ON
            outputRequestWidth = 1920
            outputRequestHeight = 1080
        }

        val contract = CropImageContractOptions(imageUri, options)
        launcher.launch(contract)
    }

    @JvmStatic
    fun registerLauncherForFragment(
        fragment: androidx.fragment.app.Fragment,
        refSendImages: StorageReference,
        linearPhotoView: LinearLayout,
        linearPhoto: LinearLayout,
        photo: ImageView,
        loadingPhoto: ProgressBar,
        loadingButton: ProgressBar,
        frameSendMsg : FrameLayout,
        msg: EditText,
        btnCamera: ImageView,
        btnSendMsg: ImageView,
        onCroppedUri: java.util.function.Consumer<Uri>
    ): ActivityResultLauncher<CropImageContractOptions> {
        return registerLauncher(
            caller = fragment,
            ctx = fragment.requireContext(),
            refSendImages = refSendImages,
            linearPhotoView = linearPhotoView,
            linearPhoto = linearPhoto,
            photo = photo,
            loadingPhoto = loadingPhoto,
            loadingButton = loadingButton,
            frameSendMsg = frameSendMsg,
            msg = msg,
            btnCamera = btnCamera,
            btnSendMsg = btnSendMsg
        ) { uri -> onCroppedUri.accept(uri) }
    }

}
